openfermioncirq/primitives/_ffft.py

```diff
@@ -0,0 +1,192 @@
+    encoded on consecutive qubits:
+    :math:`a_0^\dagger \lvert 0 \rangle = \lvert 10_2 \rangle` and
+    :math:`a_1^\dagger \lvert 0 \rangle = \vert 01_2 \rangle`.
+
+    Internally, this leads to expansion of :math:`F_0^\dagger`:
+
+    .. math::
+        \langle 0 \rvert F_0^\dagger \lvert 0 \rangle = 1
+
+    .. math::
+        \langle 01_2 \rvert F_0^\dagger \lvert 01_2 \rangle =
+            -{1 \over \sqrt{2}}
+
+    .. math::
+        \langle 10_2 \rvert F_0^\dagger \lvert 10_2 \rangle =
+        \langle 10_2 \rvert F_0^\dagger \lvert 01_2 \rangle =
+        \langle 01_2 \rvert F_0^\dagger \lvert 10_2 \rangle = {1 \over \sqrt{2}}
+
+    .. math::
+        \langle 11_2 \rvert F_0^\dagger \lvert 11_2 \rangle = -1 \, .
+    """
+
+    def __init__(self):
+        """Initializes :math:`F_0` gate."""
+        cirq.TwoQubitMatrixGate.__init__(
+            self,
+            np.array([[1,          0,         0,  0],
+                      [0, -2**(-0.5), 2**(-0.5),  0],
+                      [0,  2**(-0.5), 2**(-0.5),  0],
+                      [0,          0,         0, -1]]))
+
+    def _circuit_diagram_info_(self, args: cirq.CircuitDiagramInfoArgs
+                               ) -> cirq.CircuitDiagramInfo:
+        if args.use_unicode_characters:
+            symbols = 'F₀', 'F₀'
+        else:
+            symbols = 'F0', 'F0'
+        return cirq.CircuitDiagramInfo(wire_symbols=symbols)
+
+
+class _TwiddleGate(cirq.ZPowGate):
+    r"""Gate that introduces arbitrary FFT twiddle factors.
+
+    Realizes unitary gate :math:`\omega^{k\dagger}_n` that phases creation
+    operator :math:`a^\dagger_x` according to:
+
+    .. math::
+        \omega^{k\dagger}_n a^\dagger_x \omega^k_n =
+            e^{-2 \pi i {k \over n}} a^\dagger_x \, .
+
+    Under JWT representation this is realized by appropriately rotated pauli Z
+    gate acting on qubit x.
+    """
+    def __init__(self, k, n):
+        """Initializes Twiddle gate.
+
+        Args:
+            k: Nominator appearing in the exponent.
+            n: Denominator appearing in the exponent.
+        """
+        cirq.ZPowGate.__init__(self, exponent=-2 * k / n, global_shift=0)
+        self.k = k
+        self.n = n
+
+    def _circuit_diagram_info_(self, args: cirq.CircuitDiagramInfoArgs
+                               ) -> cirq.CircuitDiagramInfo:
+        if args.use_unicode_characters:
+            symbols = 'ω^%d_%d' % (self.k, self.n),
+        else:
+            symbols = 'w^%d_%d' % (self.k, self.n),
+        return cirq.CircuitDiagramInfo(wire_symbols=symbols)
+
+
+F0 = _F0Gate()
+
+
+def ffft(qubits: Sequence[cirq.QubitId]) -> cirq.OP_TREE:
+    r"""TODO"""
+    n = len(qubits)
+
+    if n == 0:
+        raise ValueError('Number of qubits is 0.')
+
+    if n == 1:
+        return []
+
+    if n == 2:
+        return F0(*qubits)
+
+    if n % 2 != 0:
+        raise ValueError('Number of qubits is not a power of 2.')
```

Comment:
This is not a correct check for powers of 2. It will e.g. allow the number 6 to pass.

You can use the expression `n & (n-1)` to detect non-powers-of-2. Note that this expression thinks zero is a power of two (because 2^{-infinity} = 0).

Reply:
I was counting that this check will fail at the lower recursion levels, e.g. 6 will be split to ffft of 3, which will fail on parity check.

This seems much simpler than playing with the bits? Also, if #327 is merged (which is not decided yet), then this check is going away anyway.